feature: Add ApplicationInsights telemetry

RussKie opened the pull request:
Relates to #6021

The implementation is shamelessly borrowed from https://github.com/NuGetPackageExplorer/NuGetPackageExplorer, credits to @onovotny

<!-- Please read CONTRIBUTING.md before submitting a pull request -->

## Proposed changes

* Add functionality to capture application telemetry information via AppInsights
* Add privacy statement
* Modify setup to capture user's consent to capture telemetry information
![image](https://user-images.githubusercontent.com/4403806/57295589-31c53780-710e-11e9-8d87-ddbefc58161c.png)
* Add functionality to toggle telemetry information capture in a form of a status strip button
![image](https://user-images.githubusercontent.com/4403806/57588374-a5c06f00-7556-11e9-81f2-9289167ad463.png)
* Allow configure telemetry via _Settings_
![image](https://user-images.githubusercontent.com/4403806/57301948-773d3100-711d-11e9-858e-5e94d7b92989.png)
* Prompt user to allow telemetry upon start up, if it isn't configured
![image](https://user-images.githubusercontent.com/4403806/60513985-d84b4480-9d1b-11e9-9077-0bf5cf88f5d2.png)
The text is localisable:
![image](https://user-images.githubusercontent.com/4403806/60513448-9968bf00-9d1a-11e9-8979-ccfb2310acbb.png)




The following information is captured:
* Application-level:
    * App version number (e.g. 2.0.x.x)
    * Build type (whether the application is an official release build or not)
    * Selected layout settings (such as visibility of the left panel, commit info position etc)
    * Change of selected layout settings
    * Git version (e.g. 2.19.0)
    * SSH client (e.g. OpenSSH or PuTTY)
* Operating System-Level:
    * OS version (e.g. Windows 10.0.17763.0)
    * .NET CLR version (e.g. 4.0.30319.42000)
    * .NET SDK version (e.g. dotnet:2.10.0-24102)
    * Current culture
    * Current UI culture
    * Number of monitors
    * Resolution of all monitors
    * Primary monitor DPI / scale factor



## Screenshots <!-- Remove this section if PR does not change UI -->

* `Startup.Environment`
![image](https://user-images.githubusercontent.com/4403806/57075278-b3d9e880-6d29-11e9-8188-54569b312ac0.png)

* `Startup.FormBrowse`
![image](https://user-images.githubusercontent.com/4403806/57075352-e84da480-6d29-11e9-8997-57093f29dd40.png)





----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
I have initiates a 30 day trial of AppInsights on Azure and would like to merge the current PR into master for us to collect info and see whether we wish to collect more metrics/events, whether we should name events differently, build dashboard/charts etc...

@gitextensions/git-extensions-source you can login into the Azure portal using the credentials shared in the private channel and play with the dashboards.

mstv commented on GitExtensions/Program.cs:
```diff
@@ -48,6 +49,13 @@ private static void Main()
+                //
+                Application.ApplicationExit += (s, e) => DiagnosticsClient.OnExit();
+                AppDomain.CurrentDomain.UnhandledException += (s, e) => DiagnosticsClient.Notify((Exception)e.ExceptionObject);
+                Application.ThreadException += (s, e) => DiagnosticsClient.Notify(e.Exception);
```
Nitpick: What comment did you have in mind for line 54?

RussKie replied:
Mmm... it was late in the evening... 😊 
Thanks for spotting

mstv commented on GitUI/Infrastructure/Telemetry/DiagnosticsClient.cs:
```diff
@@ -0,0 +1,110 @@
+
+                // languages
+                { "CurrentCulture", CultureInfo.CurrentCulture.Name },
+                { "CurrentUICulture", CultureInfo.CurrentUICulture.Name }
```
We should trace the `AppSettings.CurrentTranslation`, too. (I don't like localized development tools at all.)

RussKie replied:
👍

mstv commented on GitUI/Infrastructure/Telemetry/DiagnosticsClient.cs:
```diff
@@ -0,0 +1,110 @@
+            _client.Flush();
+
+            // Allow time for flushing:
+            System.Threading.Thread.Sleep(1000);
```
This is OK for me since the application still disappears at once. It is only visible in the Process Explorer for this additional second.

RussKie replied:
I should have mentioned it that I shamelessly borrows implementations from https://github.com/NuGetPackageExplorer/NuGetPackageExplorer.

mstv commented:
The portable build is missing a file.
My local debug build seems to work well.
It has created the following additional DLLs:
![DLLs](https://user-images.githubusercontent.com/36601201/57080128-3050d500-6cf2-11e9-94cf-e6adfba99c98.png)

  Problemereignisname:	CLR20r3
  Problemsignatur 01:	GitExtensions.exe
  Problemsignatur 02:	3.2.0.443
  Problemsignatur 03:	5ccae7de
  Problemsignatur 04:	GitExtensions
  Problemsignatur 05:	3.2.0.443
  Problemsignatur 06:	5ccae7de
  Problemsignatur 07:	2
  Problemsignatur 08:	72
  Problemsignatur 09:	System.IO.FileNotFoundException
  Betriebsystemversion:	6.1.7601.2.1.0.256.48
  Gebietsschema-ID:	1031
  Zusatzinformation 1:	e6e6
  Zusatzinformation 2:	e6e6c208b90db2aa028f8f653cabbf23
  Zusatzinformation 3:	dab7
  Zusatzinformation 4:	dab7e87280b076b693a79044fcd2eaee

mstv commented on GitUI/Infrastructure/Telemetry/DiagnosticsClient.cs:
```diff
@@ -0,0 +1,110 @@
+
+                // languages
+                { "CurrentCulture", CultureInfo.CurrentCulture.Name },
+                { "CurrentUICulture", CultureInfo.CurrentUICulture.Name }
```
We should trace `AppSettings.CurrentTranslation`, too. (I don't like localized development tools at all.)

mstv commented on GitUI/Infrastructure/Telemetry/DiagnosticsClient.cs:
```diff
@@ -0,0 +1,110 @@
+            _client.Flush();
+
+            // Allow time for flushing:
+            System.Threading.Thread.Sleep(1000);
```
This is OK for me since the application disappears at once. It is only visible in the Process Explorer for this additional second.

RussKie commented:
> We should have an opt-in dialog, especially for European users.

Ah yes, thank you for reminding me about your GDPR rules.
👍 

> The portable build is missing a file.

Thank you, another things I've overlooked.

mstv commented:
The portable builds are complete now. Though the second AppVeyor build - the `Required` one - references a temporary(?) merge commit, which is only known to GH and cannot be seen in GE:
- Git Extensions 3.2.0.5782 Build 71f996350b2ebd55cbd05d7023c5f4963c77ec4e (Dirty)
- Git Extensions 3.2.0.5781 Build 458c9522a281fddc4de4482646c5f65de8cc3af1 (Dirty)

RussKie commented:
> The portable builds are complete now.

Thank you for confirming it.

I'm going to add an "opt-out" switch too.

I've been thinking to add telemetry to the RC too. Thoughts?

onovotny commented on GitUI/Infrastructure/Telemetry/DiagnosticsClient.cs:
```diff
@@ -0,0 +1,113 @@
+        public static void Initialize(bool isDirty)
+        {
+            TelemetryConfiguration.Active.InstrumentationKey = "2ef275e3-8850-4305-9d7c-825a60c3d296";
+            TelemetryConfiguration.Active.TelemetryChannel.DeveloperMode = Debugger.IsAttached;
```
You no longer need this if you use the latest AppInsights.WindowsDesktop library

onovotny commented on GitUI/GitUI.csproj:
```diff
@@ -65,6 +66,9 @@
     <PackageReference Include="Microsoft-WindowsAPICodePack-Shell">
       <Version>1.1.3.3</Version>
     </PackageReference>
+    <PackageReference Include="Microsoft.ApplicationInsights.WindowsServer.TelemetryChannel">
```
I would recommend AppInsights.WindowsDesktop instead as the persistence channel is more robust. It writes to disk first and resends on next launch if needed.

RussKie replied:
Thank you for taking time to review 👍

RussKie commented:
> We should have an opt-in dialog, especially for European users.

Added a button to toggle the telemetry
![image](https://user-images.githubusercontent.com/4403806/57138644-90c93a80-6df6-11e9-9974-4740d90d2d07.png)

I think _Settings > General_ is the logic place to add the setting, however the tab is already full and requires some rethinking/rework before we can add the setting.

onovotny commented:
Couple more things you may want to consider:

The ApplicationInsights.config file has additional configuration:
https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/master/PackageExplorer/ApplicationInsights.config

You can do it in code if you want as well, but you'd need to add the same things, like the PersistenaceChannel.

You can also use the Session initializer to get a session id correlation and there's also modules/initializers for default device and unhandled exceptions.

gerhardol commented:
> I'm going to add an "opt-out" switch too.

It seems like it is an "opt-in", which is OK

> I've been thinking to add telemetry to the RC too. Thoughts?

I would prefer to have some documentation (and info popup) for inclusion in official builds. Maybe OK for a RC2, but this should maybe be removed in the final build.

 > The following information is captured:

Missing at least:
* Git version
* SSH config (primary OpenSSH or PuTTY)
* Current repo is ssh or http
* diff/mergetool

mast-eu commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -506,8 +507,37 @@ protected override void OnLoad(EventArgs e)
+
+                        // revision grid
+                        { nameof(AppSettings.ShowAuthorAvatarColumn), AppSettings.ShowAuthorAvatarColumn.ToString() },
+                        { nameof(AppSettings.ShowAuthorNameColumn), AppSettings.ShowAheadBehindData.ToString() },
```
```suggestion
                        { nameof(AppSettings.ShowAuthorNameColumn), AppSettings.ShowAuthorNameColumn.ToString() },
```

RussKie commented:
Thanks @onovotny 👍 



> > I'm going to add an "opt-out" switch too.
> 
> It seems like it is an "opt-in", which is OK

It is "opt-out" as it enabled by default.



> I would prefer to have some documentation (and info popup) for inclusion in official builds. Maybe OK for a RC2, but this should maybe be removed in the final build.

I added a privacy policy (copied from NuGetPackageExplorer, kudos to Oren).
Not sure what the best way to make the user aware of the telemetry. Ideas are welcome.




> Missing at least:
> 
> * Git version
> * SSH config (primary OpenSSH or PuTTY)
> * Current repo is ssh or http
> * diff/mergetool

Added a collection for git/ssh.
The other two are a bit tricky, can be added later (as well as many other things).

mstv commented on GitUI/Infrastructure/Telemetry/FormBrowseDiagnosticsReporter.cs:
```diff
@@ -0,0 +1,45 @@
+
+                    // revision grid
+                    { nameof(AppSettings.ShowAuthorAvatarColumn), AppSettings.ShowAuthorAvatarColumn.ToString() },
+                    { nameof(AppSettings.ShowAuthorNameColumn), AppSettings.ShowAheadBehindData.ToString() },
```
as already mentioned by @mast-eu
```suggestion
                    { nameof(AppSettings.ShowAuthorNameColumn), AppSettings.ShowAuthorNameColumn.ToString() },
```

mstv commented on PrivacyPolicy.md:
```diff
@@ -0,0 +1,50 @@
+* Exception information
+  * Could, in rare cases, contain paths packages on your computer
+* Machine name
+* Host name
```
I cannot see how names of machine and host could help tracking down anything (data minimization!).

RussKie replied:
I've looked through the logs and I don't see them logged

mstv replied:
Good. Then the regarding three lines could be removed.

mstv replied:
You have already removed these two of the three lines but kept a copy of the original section "Application-Level".

mstv commented on PrivacyPolicy.md:
```diff
@@ -0,0 +1,50 @@
+* Version (e.g. Windows 10.0.17763.0)
+* Build (e.g. 17134.1.amd64fre.rs4_release.180410-1804)
+* Available processors/cores (e.g. 8 cores)
+* Machine Name (e.g. MyFastPC)
```
ditto

RussKie replied:
It isn't something we collect explicitly, I'd think these are collected by one of packages referenced in ApplicationInsights.config.
I'm guessing, it would be `DeviceTelemetryInitializer`

 @onovotny are you able to shed some light?
Is there any way to exclude these extra props we don't require?

onovotny replied:
Those are set in the AppInsights TelemetryClient:
https://github.com/Microsoft/ApplicationInsights-dotnet/blob/80025b5d79cc52485510d422cfa5a0a8159dac83/src/Microsoft.ApplicationInsights/TelemetryClient.cs#L544

Looks like you could manually set RoleInstance to something else using a TelemetryInitalizer and it'd avoid sending it.

mstv commented on GitUI/Infrastructure/Telemetry/AppStartDiagnosticsReporter.cs:
```diff
@@ -0,0 +1,37 @@
+            }
+            else if (GitCommandHelpers.Plink())
+            {
+                sshClient = "PutTy";
```
```suggestion
                sshClient = "PuTTY";
```

mstv commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -90,6 +90,12 @@ static AppSettings()
 
+        public static bool TelemetryEnabled
+        {
+            get => GetBool("TelemetryEnabled", true);
```
We could add another Setting like `TelemetryApproved` in oder to ask the user on the first start of GE.
```suggestion
            get => GetBool("TelemetryEnabled", false);
```

RussKie replied:
I'm planning something better
![image](https://user-images.githubusercontent.com/4403806/57212988-c9505a80-7028-11e9-816d-0b1e26a88335.png)

Need to work out how to execute a custom action at the end of the install.

mast-eu replied:
I like this idea. The installer is the right place to inform about telemetry and giving the option to opt-out. In this case I also agree with having the checkbox selected by default (i.e. _opt-out_, not _opt-in_), because it is well visible.

mstv replied:
Yes, that's less intrusive than a nag screen. But the portable version will still need something else.

RussKie replied:
> But the portable version will still need something else.

Bah! Yes, you're right 😒 
Any thoughts?

RussKie replied:
I made the button more prominent, perhaps it is sufficient.
It is not visible in the dashboard though...

![image](https://user-images.githubusercontent.com/4403806/57489266-d0ad8780-72f8-11e9-8e65-d7c1b4934e9c.png)

gerhardol replied:
The default should still be off, use opt-in

RussKie replied:
Respectfully disagree, if we have it disabled by default no one will turn it on. It is just a fact of life.

However if we have it on by default, those who actively dislike tracking (I'm one of those) will go out of their way and turn if off.

RussKie commented on GitExtensions/Program.cs:
```diff
@@ -48,6 +49,17 @@ private static void Main()
+
+                // Report diagnostics irrespective of whether a debugger attached or not
+                AppDomain.CurrentDomain.UnhandledException += (s, e) => DiagnosticsClient.Notify((Exception)e.ExceptionObject);
+                Application.ThreadException += (s, e) => DiagnosticsClient.Notify(e.Exception);
```
@onovotny do I need these having the following in the ApplicationInsights.config:
```
  <TelemetryModules>
    <Add Type="Microsoft.ApplicationInsights.WindowsDesktop.DeveloperModeWithDebuggerAttachedTelemetryModule, AppInsights.WindowsDesktop"/>
    <Add Type="Microsoft.ApplicationInsights.WindowsDesktop.UnhandledExceptionTelemetryModule, AppInsights.WindowsDesktop"/>
    <Add Type="Microsoft.ApplicationInsights.WindowsDesktop.UnobservedExceptionTelemetryModule, AppInsights.WindowsDesktop" />
    <!--<Add Type="Microsoft.ApplicationInsights.WindowsDesktop.FirstChanceExceptionStatisticsTelemetryModule, AppInsights.WindowsDesktop" />-->
  </TelemetryModules>
```

I peeked inside, and they looked like they are handling the same events.

onovotny replied:
Those are handling many of the same events, it's often easier to let the initializers do it though.

vbjay commented:
Instead of registry use a per user setting. Let the registry setting act as
the default. Then portable will be saved just like all settings saved to
user settings.

On Tue, May 7, 2019, 3:25 AM Igor Velikorossov <notifications@github.com>
wrote:

> *@RussKie* commented on this pull request.
> ------------------------------
>
> In GitCommands/Settings/AppSettings.cs
> <https://github.com/gitextensions/gitextensions/pull/6522#discussion_r281502485>
> :
>
> > @@ -90,6 +90,12 @@ static AppSettings()
>
>              }
>
>          }
>
>
>
> +        public static bool TelemetryEnabled
>
> +        {
>
> +            get => GetBool("TelemetryEnabled", true);
>
>
> But the portable version will still need something else.
>
> Bah! Yes, you're right 😒
> Any thoughts?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6522#discussion_r281502485>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3WDMNQK4NPOESBYOT5HATPUEVF7ANCNFSM4HJ54ZMQ>
> .
>

RussKie commented:
> Instead of registry use a per user setting. Let the registry setting act as the default. Then portable will be saved just like all settings saved to user settings.

The telemetry setting is store in `C:\Users\<user>\AppData\Roaming\GitExtensions\GitExtensions\GitExtensions.setting`.
However I think you somewhat miss the point of the conversation.

A user's consent to capture telemetry information is captured during a setup (please refer to the screenshots in the description). The consent is then injected into the user's settings file.

The question how we can make a user aware of the telemetry in the portable version _in a nice and simple way_. 
I would like it to be "opt out" (i.e. enabled by default).

vbjay commented:
We have the in a rebase and merge conflict notifications on the bottom of
the browser window. Provide another that shows on or off with colors and
text. When clicked it let's user alter the setting.

On Tue, May 7, 2019, 7:37 AM Igor Velikorossov <notifications@github.com>
wrote:

> Instead of registry use a per user setting. Let the registry setting act
> as the default. Then portable will be saved just like all settings saved to
> user settings.
>
> The telemetry setting is store in
> C:\Users\<user>\AppData\Roaming\GitExtensions\GitExtensions\GitExtensions.setting
> .
> However I think you somewhat miss the point of the conversation.
>
> A user's consent to capture telemetry information is captured during a
> setup (please refer to the screenshots in the description). The consent is
> then injected into the user's settings file.
>
> The question how we can make a user aware of the telemetry in the portable
> version *in a nice and simple way*.
> I would like it to be "opt out" (i.e. enabled by default).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6522#issuecomment-490042995>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3WDMPT77FSI2DWBQIDZRDPUFSWLANCNFSM4HJ54ZMQ>
> .
>

RussKie commented on GitUI/Infrastructure/Telemetry/DiagnosticsClient.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            if (!Enabled)
+            {
+                return;
```
Thanks, the call in `Initialize` is meant to go via `TrackEvent()` that has the safe guard.
Fixed

RussKie commented:
> We have the in a rebase and merge conflict notifications on the bottom of the browser window. Provide another that shows on or off with colors and text. When clicked it let's user alter the setting.

That's a good idea 👍

RussKie commented:
All cleaned and made up, ready for another round of reviews.
I have updated the PR's description and added screenshots.

The only outstanding item is a better notification for a portable version. @vbjay's idea sounds good.

mstv commented on PrivacyPolicy.md:
```diff
@@ -0,0 +1,42 @@
+    * Primary monitor DPI / scale factor
+
+
+### Application-Level
```
This section is an outdated copy.

RussKie replied:
Thank you, a bad rebase at midnight...

mstv commented on PrivacyPolicy.md:
```diff
@@ -0,0 +1,42 @@
+### Application-Level
+Includes:
+* Exception information (could, in rare cases, contain paths on your computer)
+  * Version number (e.g. 2.0.x.x)
```
indent could be removed

mstv commented on PrivacyPolicy.md:
```diff
@@ -0,0 +1,42 @@
+### Operating System-Level
+Includes:
+    * Version (e.g. Windows 10.0.17763.0)
+    * Machine Name (e.g. MyFastPC)
```
Can this be removed, too?

onovotny replied:
Unless you're explicitly setting the "RoleInstance" property in an initializer, AppInsights core library will set it to the hostname. It's your choice, but if you do that, I'd include it in the disclosures.

I find OS version quite useful to record so you know what API's you can safely use.

RussKie replied:
@mstv could you elaborate why you are concerned about the hostname?

For me it seems more effort in removing it since it is captured by AppInsights libraries. 
(To a degree) I trust these libraries to make reasonable decisions about collecting a minimal subset of data.

mstv replied:
I meant only the machine name. It might contain *personal or identifiable information*, e.g. the owner's name.

RussKie replied:
Using Oren's tip, it appeared relatively painless to override the hostname capture.

RussKie commented:
Two options, need to decide which one. I don't want both of them.

1. A toolbar button (a1072fb4a687281f7e7a5b01f0e24cfbc94b2fe2)
    + more prominent
    + has look and feel of a button
    - not visible in the dashboard

![image](https://user-images.githubusercontent.com/4403806/57489266-d0ad8780-72f8-11e9-8e65-d7c1b4934e9c.png)

2. A status bar label (7e7df51e2289524ca9ce61a5bc392c61093a2a57)
    + always visible (including the dashboard)
    - doesn't have a look and feel of a button

![image](https://user-images.githubusercontent.com/4403806/57529778-16aa3000-7379-11e9-9db9-6d5e88ef34cb.png)


Thoughts? Opinions?

mast-eu commented:
I vote for the statusbar. It is common these days to have things in the statusbar that are actually clickable and not only labels. Visual Studio, VS Code, my webbrowser, FTP and email clients all have it. Even the trusty MS Office.
Thus, I think the common user can figure out that this is clickable, too.

On May 10, 2019 3:17:53 PM GMT+02:00, Igor Velikorossov <notifications@github.com> wrote:
> Two options, need to decide which one. I don't want both of them.
> 
> 1. A toolbar button (a1072fb4a687281f7e7a5b01f0e24cfbc94b2fe2)
>     + more prominent
>     + has look and feel of a button
>     - not visible in the dashboard
> 
> ![image](https://user-images.githubusercontent.com/4403806/57489266-d0ad8780-72f8-11e9-8e65-d7c1b4934e9c.png)
> 
> 2. A status bar label (7e7df51e2289524ca9ce61a5bc392c61093a2a57)
>     + always visible (including the dashboard)
>     - doesn't have a look and feel of a button
> 
> ![image](https://user-images.githubusercontent.com/4403806/57529778-16aa3000-7379-11e9-9db9-6d5e88ef34cb.png)
> 
> 
> Thoughts? Opinions?
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/gitextensions/gitextensions/pull/6522#issuecomment-491285355

gerhardol commented:
Vote for statusbar (assuming it is shown until selected once and accepted/declined).

gerhardol commented on GitUI/Infrastructure/Telemetry/AppStartDiagnosticsReporter.cs:
```diff
@@ -0,0 +1,38 @@
+            }
+            else if (GitCommandHelpers.Plink())
+            {
+                sshClient = "PuTTY";
```
This should be logging the path to find what PuTTY (bundled/external) or what external client that is used.

gerhardol replied:
> Are you able to help? 

Will look at this tonight. Maybe just internal/external putty is sufficient.

RussKie commented:
Alright, status bar it is then. Thank you.

> This should be logging the path to find what PuTTY (bundled/external) or
what external client that is used.

Are you able to help?

On Sat, May 11, 2019, 9:23 AM Gerhard Olsson <notifications@github.com>
wrote:

> *@gerhardol* commented on this pull request.
> ------------------------------
>
> In GitUI/Infrastructure/Telemetry/AppStartDiagnosticsReporter.cs
> <https://github.com/gitextensions/gitextensions/pull/6522#discussion_r283067805>
> :
>
> > +{
> +    public sealed class AppStartDiagnosticsReporter
> +    {
> +        private readonly ISshPathLocator _sshPathLocator = new SshPathLocator();
> +
> +        public void Report()
> +        {
> +            string sshClient;
> +            var sshPath = _sshPathLocator.Find(AppSettings.GitBinDir);
> +            if (string.IsNullOrEmpty(sshPath))
> +            {
> +                sshClient = "OpenSSH";
> +            }
> +            else if (GitCommandHelpers.Plink())
> +            {
> +                sshClient = "PuTTY";
>
> This should be logging the path to find what PuTTY (bundled/external) or
> what external client that is used.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6522#pullrequestreview-236345325>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXQCTQBLZ4ZTAUWFJHLPUX7X3ANCNFSM4HJ54ZMQ>
> .
>

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.Designer.cs:
```diff
@@ -221,26 +222,32 @@ private void InitializeComponent()
             // 
             toolStripMenuItem2.Name = "toolStripMenuItem2";
-            toolStripMenuItem2.Size = new System.Drawing.Size(122, 22);
+            toolStripMenuItem2.Size = new System.Drawing.Size(184, 34);
```
The commit that moves to the status bar also increases the size for icons in the button bar, looks ugly. Designer...

RussKie replied:
I'll double check but I am unable to see any artifacts locally.

gerhardol replied:
Should be
![image](https://user-images.githubusercontent.com/6248932/57588506-6cb3da80-7515-11e9-9ce5-09cbae24b4f4.png)

Is at 100%
![image](https://user-images.githubusercontent.com/6248932/57588510-7e957d80-7515-11e9-93b3-eb48d3a35d94.png)

gerhardol commented on GitUI/CommandsDialogs/SettingsDialog/Pages/GeneralSettingsPage.cs:
```diff
@@ -156,5 +138,10 @@ private void ShowGitStatus_CheckedChanged(object sender, System.EventArgs e)
+
+        private void LlblTelemetryPrivacyLink_LinkClicked(object sender, System.Windows.Forms.LinkLabelLinkClickedEventArgs e)
+        {
+            Process.Start("https://github.com/gitextensions/gitextensions/blob/master/PrivacyPolicy.md");
```
This links to master. Should we link to the release branch instead? Or the wiki?

RussKie replied:
Good point. I think linking to the branch's doc is better because it outlines what each version captures.
In set_version_to.py we update the link to the docs, in the same manner we can update the link to the branch's document:
https://github.com/gitextensions/gitextensions/blob/5e825a186d8ea793be4c3bff9893ab56b158700b/Setup/set_version_to.py#L38-L45

RussKie replied:
Mulling it over, the wiki points to [the master version for COC](https://github.com/gitextensions/gitextensions/blob/master/CODE_OF_CONDUCT.md). 
So it is probably ok to point to the privacy policy in the same fashion.

We can create a separate wiki page, if you think it is worth it, but I don't see much value in it.

I have also played with the py script, but it seems complex and brittle (with my limited knowledge of python).

gerhardol commented on GitUI/Infrastructure/Telemetry/AppStartDiagnosticsReporter.cs:
```diff
@@ -0,0 +1,38 @@
+                {
+                    { "Portable", AppSettings.IsPortable().ToString() },
+                    { "Git", GitVersion.Current.ToString() },
+                    { "SSH", sshClient },
```
Setting to sshPath will be better I believe, then we can find internal/ external PuTTY as well as use of specific Other (something to improve for Other?)

RussKie replied:
I'm ok with it.
However would we run a risk of inadvertently exposing sensitive paths?

gerhardol replied:
Still prefer the ssh path

RussKie replied:
Ok, I don't have a strong preference

RussKie replied:
Done
![image](https://user-images.githubusercontent.com/4403806/60593937-24af8680-9de7-11e9-86bd-8affc67de063.png)

gerhardol commented on PrivacyPolicy.md:
```diff
@@ -0,0 +1,31 @@
+* Selected layout settings (such as visibility of the left panel, commit info position etc)
+* Change of selected layout settings
+* Git version (e.g. 2.19.0)
+* SSH client (e.g. OpenSSH or PuTTY)
```
If SSH logging is changed:
SSH client path (e.g. OpenSSH or PuTTY)

RussKie replied:
Done
![image](https://user-images.githubusercontent.com/4403806/60593937-24af8680-9de7-11e9-86bd-8affc67de063.png)

RussKie commented:
I didn't like the statusbar placement, so I moved the option under [Help] menu:
![image](https://user-images.githubusercontent.com/4403806/58095191-cbfca380-7c15-11e9-9334-fa9243e2ccc9.png)

vbjay commented:
Bad idea. That means it will be hidden. The whole point of putting it on the statusbar is visibility.

RussKie commented:
Yes, but also clutter.

I've been thinking over the past week or so - a lot of apps employ telemetry - Visual Studio, VSCode, Resharper to name a few. None of them hide the fact, but neither display permanent notifications (I haven't seen them at least).

We don't intend to hide the fact that we will be collecting telemetry data - it will be visible on the readme, in the setup and in the app's settings; however I don't want clutter the UI with unnecessary information.

vbjay commented:
Then make sure it defaults to off and then asks the user on first run.

On Tue, May 21, 2019, 8:33 AM Igor Velikorossov <notifications@github.com>
wrote:

> Yes, but also clutter.
>
> I've been thinking over the past week or so - a lot of apps employ
> telemetry - Visual Studio, VSCode, Resharper to name a few. None of them
> hide the fact, but neither display permanent notifications (I haven't seen
> them at least).
>
> We don't intend to hide the fact that we will be collecting telemetry data
> - it will be visible on the readme, in the setup and in the app's settings;
> however I don't want clutter the UI with unnecessary information.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6522?email_source=notifications&email_token=AA3WDMIR2JKRHM5XZOKEISLPWPT27A5CNFSM4HJ54ZM2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODV3YHVQ#issuecomment-494371798>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA3WDMNULTSKH2ZSYN443T3PWPT27ANCNFSM4HJ54ZMQ>
> .
>

gerhardol commented:
> Then make sure it defaults to off and then asks the user on first run.

Or in the installer. Default should be off. (The combination of this is that the portable builds will not have telemetry though.)
To have the checkbox in Help or settings is not much different for me.

RussKie commented:
I've rolled back the alt implementation and restored the statusbar placement to make it prominently visible.

I retain my position that the telemetry should be "on" by default, otherwise we won't get any of it.
A user will have a chance to opt-out during the installation.

I still need to investigate how we can a user for a portable version (I have an idea to test).

RussKie commented:
@onovotny I wonder if you could give me some guidance.

I'd like to track general startup settings, such as _git version_, _isPortable_, _current translation_ etc.
I'm doing it before showing the first form by building a dictionary of properties and then sending it: 
```
            DiagnosticsClient.TrackEvent("AppStart",
                new Dictionary<string, string>
                {
                    { "Portable", AppSettings.IsPortable().ToString() },
                    { "Git", GitVersion.Current.ToString() },
                    { "SSH", sshClient },
                    { nameof(AppSettings.CurrentTranslation), AppSettings.CurrentTranslation },
                    { nameof(AppSettings.StartWithRecentWorkingDir), AppSettings.StartWithRecentWorkingDir.ToString() },
                });
```

Then in the portal I need to search for a custom event `AppStart` and then expand custom properties:
![image](https://user-images.githubusercontent.com/4403806/60398019-03585b80-9b97-11e9-8962-eb6f220ddb0f.png)

Is this a correct/recommended way of collecting this kind of data?
Or should I fire a series of events - one for each property?

In the end I want to have a understanding of what git version our users use, whether they prefer to install or run portable versions, use OpenSSH vs Putty etc.

RussKie commented:
I added a check when a user starts an app and telemetry isn't configured, the user will be prompted:
![image](https://user-images.githubusercontent.com/4403806/60513985-d84b4480-9d1b-11e9-9077-0bf5cf88f5d2.png)


The text is localisable:
![image](https://user-images.githubusercontent.com/4403806/60513448-9968bf00-9d1a-11e9-8979-ccfb2310acbb.png)

mstv commented:
@RussKie, do you have a clue why a FormCommit test failed?

```
1) Error : GitUITests.CommandsDialogs.CommitDialog.FormCommitTests.Should_show_committer_info_on_open
System.InvalidOperationException : Value Dispose() cannot be called while doing CreateHandle().
   at System.Windows.Forms.Control.Dispose(Boolean disposing)
   at System.Windows.Forms.ToolStrip.Dispose(Boolean disposing)
   at System.Windows.Forms.StatusStrip.Dispose(Boolean disposing)
   at System.ComponentModel.Component.Dispose()
   at System.Windows.Forms.Control.Dispose(Boolean disposing)
   at System.Windows.Forms.ContainerControl.Dispose(Boolean disposing)
   at System.Windows.Forms.Form.Dispose(Boolean disposing)
   at GitUI.CommandsDialogs.FormCommit.Dispose(Boolean disposing) in C:\projects\gitextensions-f0pvf\GitUI\CommandsDialogs\FormCommit.cs:line 364
   at System.ComponentModel.Component.Dispose()
   at GitUI.GitUICommands.<>c__DisplayClass84_0.<StartCommitDialog>g__Action|0() in C:\projects\gitextensions-f0pvf\GitUI\GitUICommands.cs:line 505
   at GitUI.GitUICommands.DoActionOnRepo(IWin32Window owner, Boolean requiresValidWorkingDir, Boolean changesRepo, EventHandler`1 preEvent, EventHandler`1 postEvent, Func`1 action) in C:\projects\gitextensions-f0pvf\GitUI\GitUICommands.cs:line 303
   at GitUI.GitUICommands.StartCommitDialog(IWin32Window owner, Boolean showOnlyWhenChanges) in C:\projects\gitextensions-f0pvf\GitUI\GitUICommands.cs:line 518
   at GitUITests.CommandsDialogs.CommitDialog.FormCommitTests.<>c__DisplayClass20_0.<RunFormTest>b__0() in C:\projects\gitextensions-f0pvf\UnitTests\GitUITests\CommandsDialogs\FormCommitTests.cs:line 303
   at GitUITests.UITest.RunForm[T](Action showDialog, Func`2 runAsync) in C:\projects\gitextensions-f0pvf\UnitTests\GitUITests\UITest.cs:line 52
   at GitUITests.CommandsDialogs.CommitDialog.FormCommitTests.RunFormTest(Func`2 testDriverAsync, CommitKind commitKind) in C:\projects\gitextensions-f0pvf\UnitTests\GitUITests\CommandsDialogs\FormCommitTests.cs:line 297
   at GitUITests.CommandsDialogs.CommitDialog.FormCommitTests.RunFormTest(Action`1 testDriver, CommitKind commitKind) in C:\projects\gitextensions-f0pvf\UnitTests\GitUITests\CommandsDialogs\FormCommitTests.cs:line 286
   at GitUITests.CommandsDialogs.CommitDialog.FormCommitTests.Should_show_committer_info_on_open() in C:\projects\gitextensions-f0pvf\UnitTests\GitUITests\CommandsDialogs\FormCommitTests.cs:line 53
```

gerhardol commented on GitUI/CommandsDialogs/FormBrowse.Designer.cs:
```diff
@@ -82,7 +82,8 @@ private void InitializeComponent()
             this.FilterToolTip = new System.Windows.Forms.ToolTip(this.components);
             this.statusStrip = new System.Windows.Forms.StatusStrip();
-            this.closeStatusBarItem = new System.Windows.Forms.ToolStripStatusLabel();
+            this.tsslblTelemetryEnabled = new System.Windows.Forms.ToolStripStatusLabel();
```
I do not want to occupy space to just show that telemetry is enabled/disabled

RussKie replied:
I'm not 100% happy with the status bar myself. And with the latest change that prompts a user to opt-in/out, I've been thinking of moving it to menus - https://github.com/gitextensions/gitextensions/pull/6522#issuecomment-494367466

gerhardol replied:
As this is set during installation (or first start) it should be enough, no need to constantly inform the user in the toolbar. (about form?)

RussKie replied:
The About form is kind of overload already, but can add to the info being copied/reported by NBug

gerhardol replied:
About would be instead of status bar no need in the report

RussKie commented:
> do you have a clue why a FormCommit test failed?

No, we'll have a look later

RussKie commented:
Version 53464567.

I have moved the telemetry settings from the status bar to the menubar under _Help_ menu - refer to https://github.com/gitextensions/gitextensions/pull/6522#issuecomment-494367466.

I have also somewhat reworked how information is recorded (no changes to _what_ is captured, only _how_ is changed). 
This is to group the info for ease of discovery:
![image](https://user-images.githubusercontent.com/4403806/60593777-cd111b00-9de6-11e9-9c72-4259cdeaa7c2.png)

RussKie commented:
> do you have a clue why a FormCommit test failed?

Must have been a glitch... Tests pass now

mstv commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -3232,5 +3248,15 @@ private void FormBrowse_DragEnter(object sender, DragEventArgs e)
+
+        private void TsmiTelemetryEnabled_Click(object sender, EventArgs e)
+        {
+            UICommands.StartGeneralSettingsDialog(this);
```
I'd prefer what I guess other user would expect, too. The only little drawback of this is that the link to the PrivacyPolicy.md is not directly available.
```suggestion
            AppSettings.TelemetryEnabled = !tsmiTelemetryEnabled.Checked;
```
BTW, the time it takes to open the settings dialog is really annoying.

RussKie replied:
I specifically don't want to do that - I want to make users aware of why we collect telemetry and we can't show the link in the menus.

Agree, the settings is a little slow (it is slower in a Debug build), but I don't expect users constantly go back and forth... So delay is probably not an end of the world here. Besides it gives user a second chance to think 😉

mstv commented on ResourceManager/Strings.cs:
```diff
@@ -53,6 +62,9 @@ public static void Reinitialize()
         public static string GitExecutableNotFound => _instance.Value._gitExecutableNotFoundText.Text;
 
+        public static string TelemetryPersmissionCaption => _instance.Value._telemetryPersmissionCaption.Text;
+        public static string TelemetryPersmissionMessage => _instance.Value._telemetryPersmissionMessage.Text;
```
Replace all `elemetryPersmission` with `elemetryPermission`.

RussKie replied:
Thank you!

RussKie commented:
> I cannot build and test the installer due to MSVS 2019.

Please make sure to install the latest Wix toolset - https://wixtoolset.org/

RussKie commented:
I've had a quick chat with @onovotny today and then had a read of GDPR, and I'm starting to get convinced that we don't need to ask permissions to collect telemetry _because_ we are not collecting any personal information.
All data we collect is about the app, and it cannot be used to identify individual users.

E.g., here's what we collect:


## Page View Properties

Property | Value
-- | -- 
Country or region | Russia 
State or province | Moscow Oblast |  
City | Odintsovo |  
Page view duration | 0 |  
Event time | 12/5/2019, 3:23:18 AM (Local time) |  
Telemetry type | pageView |  
Session Id | a8b87cad-1bf8-4b1f-9497-11021907683a |  
User Id | uDI//FwN5ESVKUIhU6g7oPaOPc7gzH1F8aLNTRVJ5Q0= |  
Application version | 3.2.0 |  
Device type | PC |  
Device model | System Product Name |  
Operating system | Windows 10.0.18362 |  
Client IP address | 0.0.0.0 |  
Cloud role name | Git Extensions |  
Cloud role instance | Git Extensions |  
SDK version | dotnet:2.10.0-24102 |  
Sample rate | 1 |  
View page name | GitUI.CommandsDialogs.FormBrowse |  
Performance | <250ms |  


## Custom Properties
Property | Value
-- | -- 
GE.StartWithRecentWorkingDir | False |  
Monitor primary resolution | 1920x1200 |  
GE.CurrentTranslation | English |  
Monitor secondary1 resolution | 1920x1200 |  
Monitor primary DPI | 96 |  
Thread culture | en-US |  
Monitor count | 3 |  
GE.IsPortable | False |  
Monitor secondary2 resolution | 1920x1200 |  
GE.IsRelease | False |  
Network type | Ethernet |  
CLR version | 4.8.4018.0 |  
UI culture | en-US |  
Time zone | Eastern Standard Time |  
SSH | PuTTY |  
Git | 2.22.0 |  
SSH.Path | C:\Program Files (x86)\GitExtensions\PuTTY\plink.exe

RussKie commented:
I'm planning to enable the telemetry by default and remove the page from the installer.

sharwell commented:
@RussKie We should eliminate unnecessary data linked to specific users before enabling telemetry by default. This includes:

* User Id
* Client IP address
* SSH.Path

onovotny commented:
UserId is already a pseuononymous hash and helps track usage data, ID that an issue?

mstv commented:
Yes, it is an issue regarding the GDPR. A pseudonymized id makes it still "identifiable data" because a hash function is nearly collision-free.

sharwell commented:
Put another way, it is not sufficient to only store hashes of information. If the stored telemetry _in combination with another data source_ can reassociate the telemetry with a particular user, then the telemetry would be considered personal data. In this case, a user could derive the User ID separately and accidentally post it via a different mechanism (e.g. a heap dump of a process), and the telemetry immediately becomes personally identifiable.

onovotny commented:
What is the proper way to correlate sessions for a given user then? Knowing the number of users using x feature is really useful. I don’t need to know the user details.

sharwell commented:
@onovotny The easiest way is to approximate it using the number of sessions using the feature. Most business decisions you would make by number of users will arrive at the same decision using number of sessions as an input.

RussKie commented:
I don't know what and how User Id is generated.

I have set User Id field to a custom string:
```cs
_client.Context.User.Id = "Git Extensions User";
```

...yet it isn't sent. Here's the payload we send:
```
{
    "name": "Microsoft.ApplicationInsights.Dev.3d09f61296404682829c3f6ac1ed3e86.PageView",
    "time": "2019-12-10T11:46:38.2079884Z",
    "iKey": "3d09f612-9640-4682-829c-3f6ac1ed3e86",
    "tags": {
        "ai.application.ver": "3.3.0",
        "ai.device.type": "PC",
        "ai.device.id": "5c18ca51-a6b6-4e6d-ab9a-cd2df8f476fd",
        "ai.device.osVersion": "Windows 10.0.18362",
        "ai.device.oemName": "Parallels Software International Inc.",
        "ai.device.model": "Parallels Virtual Platform",
        "ai.cloud.role": "Git Extensions",
        "ai.cloud.roleInstance": "Git Extensions",
        "ai.session.id": "24fb6206-6bc3-4462-b3f3-5261cf215f20",
        "ai.user.id": "W2h0uafNQgL39kLgfhzSxQeAiCMqib+KvShiiu1P37g=",        <--- there!
        "ai.internal.sdkVersion": "dotnet:2.10.0-24102",
        "ai.internal.nodeName": "MacW10"
    },
    "data": {
        "baseType": "PageViewData",
        "baseData": {
            "ver": 2,
            "name": "GitUI.CommandsDialogs.FormRemotes",
            "duration": "00:00:00",
            "properties": {
                "Monitor count": "1",
                "DeveloperMode": "true",
                "GE.IsRelease": "True",
                "CLR version": "4.8.4018.0",
                "Time zone": "AUS Eastern Standard Time",
                "UI culture": "en-GB",
                "Network type": "Ethernet",
                "Thread culture": "en-GB",
                "GE.StartWithRecentWorkingDir": "False",
                "Monitor primary DPI": "144",
                "Monitor primary resolution": "2868x1650",
                "GE.IsPortable": "False",
                "GE.Environment": "development",
                "SSH": "OpenSSH",
                "GE.CurrentTranslation": "English",
                "Git": "2.16.1.windows.4"
            }
        }
    }
}
```

@onovotny any thoughts?

onovotny commented:
Not sure without digging into the App Insights core code

RussKie commented:
> Most business decisions you would make by number of users will arrive at the same decision using number of sessions as an input.

I don't think I completely agree with this statement.
Right now our telemetry has records for 430K sessions from some 2,800 opted-in users for the last 90 days. Here are OS stats:

OS | Count | Percentage
--|---|--
Windows 10.0.18363 | 163425 | 37.93%
Windows 10.0.19041 | 78891 | 18.31%
Windows 10.0.19042 | 65888 | 15.29%
Windows 10.0.17763 | 52378 | 12.16%
Windows 10.0.16299 | 14021 | 3.25%
Windows 10.0.18362 | 12057 | 2.80%
Windows 10.0.17134 | 11223 | 2.60%
Windows 6.2.9200 | 9225 | 2.14%
Windows 6.1.7601 | 8959 | 2.08%
Windows 10.0.14393 | 7314 | 1.70%
Windows 6.3.9600 | 4834 | 1.12%
Windows 10.0.(other) | 2655 | 0.61%

Based on number of sessions alone one could make very different assumptions about our user base, e.g. that majority of our users are on W10.
However given that is only ~1.5% of our users (2.8K out of ~200K, based on GitHub download counter) it is a totally wild, and likely incorrect, guess.
